[Sol] Step3-02-1 사각형 뷰 생성

Hansolkkim opened the pull request:
## 실행 화면
![SS 2022-03-04 AM 01 05 40](https://user-images.githubusercontent.com/92504186/156603786-9b9c9069-eaf9-4129-94f1-8a1fee41f5b1.gif)

## 작업목록
- [x] Refatoring
  - [x] BackgroundColor, Alpha 타입 초기화 메서드 수정
  - [x] 랜덤 ID를 생성하는 위치를 ID타입 -> RectangleFactory로 변경
- [x] Plane 객체 선언
- [x] ViewFactory 객체 선언
- [x] Delegate 패턴을 이용해, 새로운 랜덤Rectangle 객체가 생성되면 ViewFactory에서 Rectangle객체로 RectangleView 생성
- [x] generateRandomRectangleButton의 동작 이벤트 생성
- [x] 배경색, 투명도 View를 Storyboard를 이용해 생성
- [x] 랜덤한 Rectangle 객체가 그려질 수 있는 drawableAreaView 구현
---
## 학습 키워드

- `frame` , `UIAction` , `view 생성`  , `정규식 표현`
---
## 고민과 해결
1. `Alpha.init(opacityLevel:)` 를 보면 어떤 값을 넣을지 불명확하다는 리뷰를 받았습니다.
  이를 해결하기 위해 Alpha의 단계에 대한 제한을 기본적으로는 (1\~10)으로 두되, 사용자가 해당 단계를 수정할 수도 있도록 Alpha 타입의 init 메서드 매개변수에 max, min 을 추가했습니다. 그리고 실제 Alpha 값의 범위는 (0\~1.0)이므로, 축척(scale)을 이용해 원하는 level을 0\~1.0 사이의 값으로 지정해주도록 했습니다.
2. BackgroundColor의 colorValue 범위가 어디부터 어디까지인지 외부에서 예측 가능한지 어렵지 않겠냐는 리뷰를 받았습니다. 
  이를 해결하기 위해 위의 Alpha와 같은 방식으로 초기화 단계에서 외부에서 BackgroundColor의 min, max를 설정해주고 원하는 값을 넣을 수 있도록 변경했습니다.
3. 랜덤 ID를 생성하는 위치가 ID 타입 내부에 있어야 되는지, 생각을 해보라는 리뷰를 받았습니다.
  나중에 만약 앱 이용자가 ID, Size, Point 등을 모두 직접 입력해서 Rectangle을 생성하게 되는 경우도 생길 수 있고, 이런 여러 상황에서도 제가 작성한 코드를 사용할 수 있도록 유연하게 설계하려면, 무조건 ID 타입은 랜덤한 값을 가지는게 맞지 않겠다는 생각이 들었고, RectangleFactory에서 랜덤 Rectangle 객체를 생성할 때에만 랜덤 ID를 생성하도록, 랜덤 ID를 생성하는 위치를 RectangleFactory 클래스로 옮겼습니다.
4. Rectangle 객체를 이용하여 사각형 View를 만드는 과정에서, Rectangle 객체의 프로퍼티들은 모두 `private` 접근 제어자를 가지고 있어 접근에 어려움이 있었습니다. 사각형 객체에서 이런 프로퍼티 값들에 대한 판단을 외부에서 지시하는 등의 상황은 존재하지 않을 것 같았고, getter를 이용해 제공해주는 것 보다는, 그냥 외부에서 해당 객체의 프로퍼티에 접근은 가능하되 수정은 불가능하도록 접근 제어자를 `private(set)` 으로 변경했습니다.

Hansolkkim committed:
Refactor: BackgroundColor, Alpha 타입 초기화 메서드 수정

타입이 어떤 "단계"를 의미한다고 가정하고, 최저단계, 최고단계를 초기화 메서드를 이용해 지정해줄 수 있도록 설정
입력받은 최저, 최대 단계를 기준으로 축척을 만들어 입력된 단계에서의 값을 계산하여 프로퍼티에 입력하도록 설정

Hansolkkim committed:
Docs: README.md 파일에 리팩토링 내용 추가

Hansolkkim committed:
Refactor: ID 타입 초기화 시에 매개변수로 String을 받도록 수정

그에 따른 초기화 메서드 수정
ID format이 정해진 제한에 맞는지 검증하는 isCorrectFormat() 메서드 ID 구조체 내에 선언

Hansolkkim committed:
Feat: Plane 객체 선언

해당 객체는 생성된 Rectangle객체를 보관
요구사항에 맞는 기능을 제공하도록 설정
Plane의 isRectangleExist(in:) 메서드를 위해 Rectangle 클래스에 hasSamePoint(_:) 메서드 선언
Rectangle클래스의 hasSamePoint(_:) 메서드를 위해 Point 타입 클래스는 Equatable 프로토콜 채택

Hansolkkim committed:
Feat: RectagleDelegate 프로토콜의 기능 추가

구현해야하는 메서드 명 factoryDidGenerateRandomRectangle(_:) 로 변경
Plane 객체에게 Rectangle 객체를 추가하라고 지시하는 기능 추가

Hansolkkim committed:
Feat: ViewController 클래스에 버튼 생성

ViewController 클래스에 addGenerateRectangleButton() 메서드 추가
ViewController 클래스 프로퍼티로 factory, generateRectangleButton 추가

Hansolkkim committed:
Feat: generateRectangle 버튼이 touch된 이벤트를 RectangleFactory에 전달하는 기능 추가

해당 동작을 위한 generateRectangleButtonTouched() 메서드 선언

Hansolkkim committed:
Refactor: Alpha 클래스가 CustomStringConvertible 프로토콜 채택

os_log로 해당 부분을 잘 알아볼 수 있도록 description 연산프로퍼티 선언

Hansolkkim committed:
Refactor: generateButton의 동작 방식 변경

UIAction을 이용해 generateButton의 이벤트 동작 설정해주는 방식으로 변경

Hansolkkim committed:
Feat: 새로운 View를 생성해주는 ViewFactory 객체 구현

해당 객체 내에 static function으로 generateRectangleView(of:) 메서드 선언
매개변수로 받은 Rectangle 객체를 생성하도록 설정
Rectangle 객체의 프로퍼티를 사용하기 위한 메서드를 만들기에는 Rectangle에게 판단하도록 하는 내용이 없는 것으로 판단되어, Rectangle 객체의 프로퍼티들의 접근 제어자 수정

Hansolkkim committed:
Style: ViewController 클래스의 프로퍼티 factory 이름 변경 및 Plane.swift 파일 불필요한 코드 제거

ViewFactory 객체의 프로퍼티가 생긴다면 헷갈릴 것이라 판단하여 factory 프로퍼티 이름을 rectangleFactory로 변경

Hansolkkim committed:
Feat: 화면 우측 statusView와 그 안에 배경색, 투명도 View들 추가

화면에 그림이 그려질 수 있는 View의 크기를 설정할때 사용하기 위해 statusView의 아울렛 변수 선언

Hansolkkim committed:
Feat: ViewController의 factoryDidGenerateRandomRectangle(_:) 메소드에 새로운 View를 생성하는 기능 추가

Hansolkkim committed:
Feat: 그림이 그려질 수 있는 범위를 나타내는 drawableAreaView 추가

그에 따라 RectangleFactory에 넘겨줄 screenSize 수정(그릴 수 있는 범위 제한)
drawableAreaView의 크기를 설정할 때 사용하기 위해 ViewController에 generateRectangleButton 프로퍼티 추가

Hansolkkim committed:
Fix: generateRandomRectangleButton의 X,Y 좌표 수정

Storyboard에 View들을 추가함으로 인해 self.view.safeAreaLayoutGuide가 변경되어, 원하는 값으로 button의 X,Y 좌표 수정

Hansolkkim committed:
Refactor: generateRectangleButton, drawableAreaView 프로퍼티의 타입 수정

두 프로퍼티는 UIView 타입의 프로퍼티이므로 타입을 ? -> ! 로 수정

godrm commented on DrawingApp/DrawingApp/Model/Alpha.swift:
```diff
@@ -10,19 +10,23 @@ import Foundation
-        } else if opacityLevel < minOpacityLevel {
-            self.value = 0
+        } else if opacityLevel < min {
+            self.value = 0.1
```
`min: Int = 1, max: Int = 10`가 바뀔 수 있는 거라면 
`Alpha.init(opacityLevel: 1, min: 0, max: 1)` 와 `Alpha.init(opacityLevel: 1, min: 1, max: 1000)` 
두 값이 같은 alpha 값이라고 할 수 있을까요?
서로 다른 값이 될 것 같은데요?

Hansolkkim replied:
> `min: Int = 1, max: Int = 10`가 바뀔 수 있는 거라면
> `Alpha.init(opacityLevel: 1, min: 0, max: 1)` 와 `Alpha.init(opacityLevel: 1, min: 1, max: 1000)`
> 두 값이 같은 alpha 값이라고 할 수 있을까요?
> 서로 다른 값이 될 것 같은데요?

비교할 때 어려움이 있겠네요.. 
해당 부분 수정하기 위해 Alpha 구조체에 `static let possibleOpacityLevels = 1...10` 이라는 프로퍼티를 선언해 외부에서 opacityLevel의 범위를 알 수 있도록 고쳐보겠습니다.

godrm commented on DrawingApp/DrawingApp/Model/BackgroundColor.swift:
```diff
@@ -8,19 +8,23 @@
-        let maxColorValue = 255
-        let minColorValue = 0
-        var colors = [r, g, b]
+    init(r: Int, g: Int, b: Int, min: Int = 0, max: Int = 255) {
```
범위를 지정하는 것은 BackgroundColor 의 책임일 것 같은데 매개변수로 처리하면 유연해지면서 전혀 다른 타입이 될 것 같습니다 
Alpha도 언급했었는데, 만약 두 값을 비교해야 한다면 같은 값이 될지 안될지 모르겠네요

Hansolkkim replied:
위에서 답변 드린 내용처럼, BackgroundColor 구조체에도 `static let possibleColorLevels = 0...255` 라는 프로퍼티 선언해 외부에서도 해당 값을 범위를 알 수 있고, 두 값을 비교하는 경우에도 쉽게 비교가 가능하도록 수정하겠습니다.

godrm commented on DrawingApp/DrawingApp/Model/Rectangle.swift:
```diff
@@ -9,26 +9,30 @@ import Foundation
+    private(set) var size: Size
+    private(set) var point: Point
+    private(set) var backgroundColor: BackgroundColor
+    private(set) var alpha: Alpha
```
private으로 작업하려고 해도 속성을 가져갈 수 있는 경우도 꽤 있을 겁니다. 그럴 때는 private(set)을 사용해도 됩니다. 
하지만 늘 정말 private(set)으로 열어야 하나. 필요한게 맞나 계속 고민해보시면 됩니다.

Hansolkkim replied:
이번 스텝에서는 객체에서 해당 프로퍼티로 무언가를 판단하는것 보다는 그냥 해당 개체의 프로퍼티를 가지고 View를 만들 때 사용하는 단순한 타입이라고 생각되어 접근제어자를 이와같이 수정했습니다.
접근 제어자를 지정할 때 항상 고민하면서 지정하도록 하겠습니다. 감사합니다!

godrm commented on DrawingApp/DrawingApp/Model/RectangleFactory.swift:
```diff
@@ -15,15 +15,16 @@ class RectangleFactory {
-    public func generateRandomRectangle() -> Rectangle {
+    public func generateRandomRectangle() -> Rectangle? {
         let size = Size(width: 150, height: 120)
         let point = generateRandomPoint()
         let backgroundColor = generateRandomColor()
         let alpha = generateRandomAlpha()
+        guard let id = generateRandomID() else { return nil }
         
-        let newRectangle = Rectangle(size: size, point: point, backgroundColor: backgroundColor, alpha: alpha)
+        let newRectangle = Rectangle(id: id, size: size, point: point, backgroundColor: backgroundColor, alpha: alpha)
         
-        delegate?.printLog(of: newRectangle)
+        delegate?.factoryDidGenerateRandomRectangle(newRectangle)
         
         return newRectangle
```
팩토리는 인스턴스를 두거나 복잡하게 만들거나 델리게이트를 갖거나 그러지 않는 경우가 많습니다. 
왜냐하면 특정한 타입의 인스턴스를 만드는 게 집중하기 때문에 static 메소드로 끝나는 편입니다
그래서 팩토리로 생성하는 경우는 옵셔널도 잘 없어요. 어떤 객체를 생성할지 잘 알고, 그대로 만들기 때문이죠
뭔가 로그를 남기는 의도는 알겠는데, 좀 과한 것 같기도 하네요. (델리게이트 부분은 그냥 두셔도 괜찮습니다)

Hansolkkim replied:
> 팩토리는 인스턴스를 두거나 복잡하게 만들거나 델리게이트를 갖거나 그러지 않는 경우가 많습니다.
> 왜냐하면 특정한 타입의 인스턴스를 만드는 게 집중하기 때문에 static 메소드로 끝나는 편입니다
> 그래서 팩토리로 생성하는 경우는 옵셔널도 잘 없어요. 어떤 객체를 생성할지 잘 알고, 그대로 만들기 때문이죠
> 뭔가 로그를 남기는 의도는 알겠는데, 좀 과한 것 같기도 하네요. (델리게이트 부분은 그냥 두셔도 괜찮습니다)

몰랐던 사실이네요,, RectangleFactory에 있는 메서드를 static 메서드로 변경하고, 델리게이트 부분을 Plane 클래스에서 사용하도록 수정해보겠습니다.

godrm commented on DrawingApp/DrawingApp/Model/RectangleFactory.swift:
```diff
@@ -37,14 +38,32 @@ class RectangleFactory {
+        }
+        
+        let randomIDParts = [generateRandomString(length: 3), generateRandomString(length: 3), generateRandomString(length: 3)]
+        return ID.init(with: randomIDParts.joined(separator: "-")) ?? nil
```
ID 타입이라면 ID를 생성하는 규칙 정도는 내부의 역할이 아닐까요? 외부에서 랜덤으로 생성해서 넘겨준다면 ID 타입은 그냥 문자열과 다름이 없지 않을까요?

Hansolkkim replied:
> ID 타입이라면 ID를 생성하는 규칙 정도는 내부의 역할이 아닐까요? 외부에서 랜덤으로 생성해서 넘겨준다면 ID 타입은 그냥 문자열과 다름이 없지 않을까요?

이전 리뷰에서 랜덤 ID를 생성해주는 위치를 생각해보라는 말씀을 듣고, 좀 더 유연한 구조를 위해 랜덤한 ID의 경우에는 외부에서만 생성해주고 ID 객체를 직접 생성할 일(사용자가 직접 아이디를 입력하는)이 생기면 문자열을 이용해 생성할 수 있도록 하려 했는데, 제가 사용한 방법으로는 ID 타입의 의미가 사라져버리는 것 같네요.. 해당 내용을 ID 객체 초기화 방법 중 하나로 만들고, 앞서 말한 직접 아이디를 입력하는 경우에도 ID 타입은 초기화될 수 있도록 수정해보겠습니다.

godrm commented on DrawingApp/DrawingApp/Model/ViewFactory.swift:
```diff
@@ -0,0 +1,33 @@
+        return newView
+    }
+    
+    private static func convertColorValueToCGFloat(r: Double, g: Double, b: Double) -> (r: CGFloat, g: CGFloat, b: CGFloat) {
+        
+        let convertedR = r/Double(255)
+        let convertedG = g/Double(255)
+        let convertedB = b/Double(255)
+        
+        return (convertedR, convertedG, convertedB)
+    }
```
이런 거라면 그냥 UIColor를 생성해주는 게 더 낫지 않을까요?

Hansolkkim replied:
> 이런 거라면 그냥 UIColor를 생성해주는 게 더 낫지 않을까요?

그런것 같습니다.. 불필요하게 여러번 값들을 수정해주었네요.. 해당 메서드 리턴 타입 UIColor로 수정하도록 하겠습니다.

godrm commented on DrawingApp/DrawingApp/Model/id.swift:
```diff
@@ -10,15 +10,21 @@ import Foundation
     
-    private static func generateRandomPartOfID(length: Int) -> String {
-        return String(NSUUID().uuidString.prefix(length))
+    private static func isCorrectFormat(_ id: String) -> Bool {
+        let correctFormat = "^[a-zA-Z0-9]..[-][a-zA-Z0-9]..[-][a-zA-Z0-9].."
+        if let _ = id.range(of: correctFormat, options: .regularExpression) {
+            return true
+        } else {
+            return false
+        }
     }
```
이런 좀 의아한 부분인데요, 생성은 외부에서 하는 데 포맷이 맞는지 판단은 내부에서 하는건가요?
둘 다 외부에서 하거나 둘 다 내부에서 하는게 짝이 맞는게 아닐까요?

Hansolkkim replied:
> 이런 좀 의아한 부분인데요, 생성은 외부에서 하는 데 포맷이 맞는지 판단은 내부에서 하는건가요?
> 둘 다 외부에서 하거나 둘 다 내부에서 하는게 짝이 맞는게 아닐까요?

위의 리뷰와 함께 참고하여 수정해보겠습니다.

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -9,25 +9,66 @@ import UIKit
+        let drawableAreaSize = (width: Double(self.drawableAreaView.frame.width),
+                                height: Double(self.drawableAreaView.frame.height))
+        
+        let rectangleFactory = RectangleFactory(screenSize: drawableAreaSize)
+        self.rectangleFactory = rectangleFactory
```
이 정도는 그냥 `self.rectangleFactory = RectangleFactory(screenSize: drawableAreaSize)`로 해도 되는건 아닐까요?

Hansolkkim replied:
> 이 정도는 그냥 `self.rectangleFactory = RectangleFactory(screenSize: drawableAreaSize)`로 해도 되는건 아닐까요?

ViewController 클래스 프로퍼티로 rectangleFactory를 선언한 것이 좀 시간이 지난 후여서 해당 부분을 수정하지않고 쓸데없는 코드를 추가한 것 같습니다. 수정하도록 하겠습니다.

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -9,25 +9,66 @@ import UIKit
+    }
+    
+    func addGenerateRectangleButton() {
+        let buttonUIAction = UIAction { _ in
+            let _ = self.rectangleFactory?.generateRandomRectangle()
+        }
+        let generateButton = UIButton(type: .custom, primaryAction: buttonUIAction)
```
UIAction을 연결하는 부분은 좋습니다.
여기서는 팩토리를 생성하는 역할이 아니라 `Plane에게 사각형을 추가해` 라고 알려줘야 하지 않을까요?
self.rectangleFactory?.generateRandomRectangle() 이 부분을 Plane에서 할 일 같아서요 
심지어 생성한 이후에 처리가 없네요. 어떻게 된걸까요? 
설마 델리게이트가 로그가 아니라 생성된 것을 알도록 하는 거라면 흐름이 이상한 것 같습니다 
여기서 할 일이 Plane에게 알려줘야 하는 겁니다. 
뷰 -> 뷰컨트롤러 -> 모델 흐름이 되어야 합니다

Hansolkkim replied:
> UIAction을 연결하는 부분은 좋습니다.
> 여기서는 팩토리를 생성하는 역할이 아니라 `Plane에게 사각형을 추가해` 라고 알려줘야 하지 않을까요?
> self.rectangleFactory?.generateRandomRectangle() 이 부분을 Plane에서 할 일 같아서요
> 심지어 생성한 이후에 처리가 없네요. 어떻게 된걸까요?
> 설마 델리게이트가 로그가 아니라 생성된 것을 알도록 하는 거라면 흐름이 이상한 것 같습니다
> 여기서 할 일이 Plane에게 알려줘야 하는 겁니다.
> 뷰 -> 뷰컨트롤러 -> 모델 흐름이 되어야 합니다

처음 Delegate를 테스트해볼때 팩토리 메서드 내에서 처리했던거에 빠져서 잘못된 흐름으로 동작하도록 한 것 같습니다. 해당 부분과 아래에서 말씀하신 부분을 토대로, 뷰컨트롤러는 Plane에게 지시를 내리고, Plane에서 새로운 Rectangle 객체를 생성하는 흐름으로 수정하도록 하겠습니다.

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -9,25 +9,66 @@ import UIKit
-    func printLog(of rectangle: Rectangle) {
+    func factoryDidGenerateRandomRectangle(_ rectangle: Rectangle) {
         os_log("\(rectangle)")
+        plane.add(rectangle: rectangle)
+        
+        let newRectangleView = ViewFactory.generateRectangleView(of: rectangle)
+        self.drawableAreaView.addSubview(newRectangleView)
```
그렇군요. 델리게이트가 이런 역할을 하려고 한거군요. 이걸 보니 확실히 알겠네요. 
팩토리에 델리게이트는 좋은 구조가 아닙니다. Plane이 핵심 모델이어야 합니다.

Hansolkkim replied:
> 그렇군요. 델리게이트가 이런 역할을 하려고 한거군요. 이걸 보니 확실히 알겠네요.
> 팩토리에 델리게이트는 좋은 구조가 아닙니다. Plane이 핵심 모델이어야 합니다.

아직 델리게이트에 대한 개념이 부족했던것 같습니다. 핵심 모델을 기준으로 델리게이트가 동작하도록 수정해보겠습니다 감사합니다!

godrm reviewed (changes requested):
수고하셨습니다. 액션에 대한 Plane 모델로 흐름을 개선해보면 좋겠습니다.

Hansolkkim committed:
Refactor: RectangleFactory 클래스의 메서드들을 static 메서드로 변경

Hansolkkim committed:
Refactor: ViewFactory클래스의 convertColorValueToCGFloat 메서드 수정

리턴 타입을 UIColor로 바꿈에 따라 메서드 이름 convertColorValueToUIColor로 수정
잘못 사용하고 있던 Rectangle 객체의 alpha 프로퍼티를 원하는 값으로 사용하도록 수정

Hansolkkim committed:
Refactor: generateButton의 UIAction 기능의 흐름 변경

RectangleDelegate 프로토콜 이름을 PlaneProtocol로 변경
채택시 구현해야 하는 메서드 이름 planeDidAddRectangle로 수정
위임하는 주체를 RectangleFactory에서 Plane으로 변경
generateButton을 터치 시, Plane 객체에게 지시하도록 변경
Plane 객체 내에서 RectangleFactory 이용해 새로운 Rectangle 객체 생성
해당 함수가 호출되면 delegate(ViewController)에게 알려, ViewFactory로 새로운 사각형 뷰를 생성하도록 지시하는 흐름으로 수정

Hansolkkim committed:
Refactor: 랜덤한 ID를 생성하는 위치를 ID 클래스 내부로 이동

ID 클래스에서 랜덤한 아이디를 생성, 검증하도록 역할 확장

Hansolkkim committed:
Refactor: Alpha 타입의 범위 지정

Hansolkkim committed:
Refactor: BackgroundColor 타입의 프로퍼티 값의 범위 지정

Hansolkkim commented:
리뷰 받았던 내용들 토대로 수정 완료했습니다!
제일 중요한 앱 동작 흐름 부분이 미흡해서 이 부분 중점적으로 수정했습니다. VC에서는 Plane과 ViewFactory 에게 지시를 내리도록 했고, Plane은 Rectangle 모델 인스턴스들을 관리하는 역할로 수정했습니다.
그리고 좀 적절치 않았던 Delegate는 Plane을 위주로 동작하도록 수정했습니다.

godrm commented on DrawingApp/DrawingApp/Model/Alpha.swift:
```diff
@@ -9,17 +9,18 @@ import Foundation
+        let min = Alpha.possibleOpacityLevels.min() ?? 1
+        let max = Alpha.possibleOpacityLevels.max() ?? 10
+        
         if opacityLevel > max {
             self.value = 1.0
         } else if opacityLevel < min {
             self.value = 0.1
         } else {
```
타입을 생성할 때 어떤 식으로 구현할지를 선택해야 하는데,  
• 범위를 벗어난 값이 들어와도 범위로 잘라서 만들어 줄 것인가  
• 혹은 아예 만들어지지 않도록 예외처리를 할 것인가  
• 아예 넣지도 못하도록 할 것인가  
여러가지 방법이 있습니다. 그 중에서도 개발자들이 실수를 줄일 수 있고, 런타임에 실행하지 않고 컴파일 타임에 점검할 수 있으면 가장 좋습니다. 
여기서는 어떻게 처리하고 있는걸까요? 이걸 사용하는 다른 사람이 어떤 부분을 기대하고 실수하고 확인하게 될까요? 그걸 생각해보세요

Hansolkkim replied:
> 타입을 생성할 때 어떤 식으로 구현할지를 선택해야 하는데,
> • 범위를 벗어난 값이 들어와도 범위로 잘라서 만들어 줄 것인가
> • 혹은 아예 만들어지지 않도록 예외처리를 할 것인가
> • 아예 넣지도 못하도록 할 것인가
> 여러가지 방법이 있습니다. 그 중에서도 개발자들이 실수를 줄일 수 있고, 런타임에 실행하지 않고 컴파일 타임에 점검할 수 있으면 가장 좋습니다.
> 여기서는 어떻게 처리하고 있는걸까요? 이걸 사용하는 다른 사람이 어떤 부분을 기대하고 실수하고 확인하게 될까요? 그걸 생각해보세요

저번 step1 리뷰에서 이 고민을 해보라는 리뷰를 받아서, 저는 범위를 벗어난 값이 들어와도 범위 내로 잘라서 만들어주는 방법을 선택했습니다. 그런데 해당 방법을 사용하면, 컴파일 타임에서 점검할 내용이 딱히 없게 되네요.. 조금 더 생각해보고 수정해보겠습니다!

godrm replied:
그래서 다시 남겨봤습니다. 이걸 메소드를 호출하는 사람은 무엇을 기대하고 호출할까요? 그걸 어떻게 확인할 수 있을까요?
실행해보지 않고도 실수했는지 안했는지 판단할 수 있을까... 그걸 표현할 수 있으면 가장 좋습니다.

godrm commented on DrawingApp/DrawingApp/Model/Plane.swift:
```diff
@@ -23,7 +24,9 @@ class Plane {
     
-    public func add(rectangle: Rectangle) {
-        rectangles.append(rectangle)
+    public func addNewRectangle(in frame: (Double, Double)) {
```
`frame: (Double, Double)` 이 부분이 의도가 좀 불명확해 보입니다. 
튜플을 쓰더라도 label을 붙여주면 좋습니다

Hansolkkim replied:
> `frame: (Double, Double)` 이 부분이 의도가 좀 불명확해 보입니다.
> 튜플을 쓰더라도 label을 붙여주면 좋습니다

아..저는 해당 값을 사용하는 위치(RectangleFactory클래스의 generateRandomPoint(in:) 메서드)에서만 해당 값을 의미있게 받아 사용하면 된다 생각해서 그 위치에서만 튜플에  label을 붙였습니다.. 그런데 생각해보니, 저 말고는 저 튜플의 순서를 알기 어렵네요. 해당 부분 label 명시해서 수정하도록 하겠습니다!
자세한 코드리뷰 감사합니다!!

godrm commented:
수고하셨습니다. 다음 단계로 넘어가봅시다!
